extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -302,26 +308,19 @@ private ShuffleSpec findShuffleSpecForNextWindow(List<OperatorFactory> operatorF
       }
     }
 
-    Map<String, ColumnWithDirection.Direction> sortColumnsMap = new HashMap<>();
-    if (sort != null) {
-      for (ColumnWithDirection sortColumn : sort.getSortColumns()) {
-        sortColumnsMap.put(sortColumn.getColumn(), sortColumn.getDirection());
-      }
-    }
-
-    if (partition == null || partition.getPartitionColumns().isEmpty()) {
-      // If operatorFactories doesn't have any partitioning factory, then we should keep the shuffle spec from previous stage.
-      // This indicates that we already have the data partitioned correctly, and hence we don't need to do any shuffling.
+    if (partition == null || partition.getPartitionColumns().isEmpty() || sort == null || sort.getSortColumns().isEmpty()) {
+      // If operatorFactories doesn't have any partitioning or sorting factory, then we should keep the shuffle spec from previous stage.
+      // This indicates that we already have the data partitioned and sorted correctly, and hence we don't need to do any shuffling.
       return null;
     }
 
     List<KeyColumn> keyColsOfWindow = new ArrayList<>();
-    for (String partitionColumn : partition.getPartitionColumns()) {
+    for (ColumnWithDirection sortColumn : sort.getSortColumns()) {
       KeyColumn kc;
-      if (sortColumnsMap.get(partitionColumn) == ColumnWithDirection.Direction.DESC) {
-        kc = new KeyColumn(partitionColumn, KeyOrder.DESCENDING);
+      if (sortColumn.getDirection() == ColumnWithDirection.Direction.DESC) {
+        kc = new KeyColumn(sortColumn.getColumn(), KeyOrder.DESCENDING);
       } else {
-        kc = new KeyColumn(partitionColumn, KeyOrder.ASCENDING);
+        kc = new KeyColumn(sortColumn.getColumn(), KeyOrder.ASCENDING);
       }
       keyColsOfWindow.add(kc);
     }
```

Comment:
seems to me that the data will be `ClusterBy` on `keyColsOfWindow` which is based on the `sortColumns` ; wouldn't that mean that different values of the sorted values may land on different workers?

its so hard to work with this without seeing the plan...I don't even know which class is the api to configure sort/partitioning in msq in-between stages - it seems to me  that would be `ShuffleSpec`? 

from what I see for results for `SortOperatorFactory` in the `WindowOperatorQueryKit` class; I think you may not remove any sort operators; as that's not handled correctly in-between stages

I believe the old approach to set this according to the `partitionColumns` was the right approach

Reply:
@kgyrtkirk The thing is that Sort operator (passed by Windowing.java to MSQ) already contains the `partition by` column.